src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -128,330 +147,392 @@ impl<T: 'static> EventLoop<T> {
         Ok(event_loop)
     }
 
-    pub fn run<F>(mut self, callback: F) -> !
+    pub fn run_ondemand<F>(&mut self, mut event_handler: F) -> Result<(), ExternalError>
     where
-        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow) + 'static,
+        F: FnMut(event::Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
     {
-        let exit_code = self.run_return(callback);
-        process::exit(exit_code);
+        if self.loop_running {
+            return Err(ExternalError::AlreadyRunning);
+        }
+
+        loop {
+            if let PumpStatus::Exit(code) = self.pump_events_with_timeout(None, &mut event_handler)
+            {
+                if code == 0 {
+                    break Ok(());
+                } else {
+                    break Err(ExternalError::ExitFailure(code));
+                }
+            }
+        }
     }
 
-    pub fn run_return<F>(&mut self, mut callback: F) -> i32
+    pub fn pump_events<F>(&mut self, event_handler: F) -> PumpStatus
     where
-        F: FnMut(Event<'_, T>, &RootEventLoopWindowTarget<T>, &mut ControlFlow),
+        F: FnMut(event::Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow),
```

Comment:
We already have imports like `RootEventLoopWindowTarget` and `Event` is already imported, so use direct imports instead.